update paper auth list

casperdcl opened the pull request:
# misc

- update spelling (closes #907, closes #910)
- update impact statistics (closes #911)

# authors

Author names automatically generated from `git` and/or `.mailmap` and ordered using `git fame -wMC` on latest master (80be780a).

- [x] check if need to reorder
  + stick with current `git fame -wMC`
  + or based on [latest Zenodo](https://zenodo.org/record/3675353)
- fixes #887 (specifically https://github.com/tqdm/tqdm/issues/887#issuecomment-588249265)

## explicit consent

Numbers mean order from `git fame -wMC`.

Unticked means no reply. ~~@lrq3000: I think we should not add them as co-authors if they did not give explicit consent, but in Acknowledgements it would be fine.~~ @lrq3000: Ok to add by default, to be inclusive per the initial consensus.

- [ ] @AbysmalBiscuit
- [ ] @charlienewey
- [ ] @chengs
- [ ] @CrazyPython
- [ ] @danielpanteleit
- [ ] @dyno <- assumed Dyno Fu
- [ ] @fordhurley
- [ ] @geekypandey
- [ ] @gsganden
- [ ] @JackMc
- [ ] @littlezz <- no full name
- [ ] @martinzugnoni
- [ ] @minho42
- [x] @noamraph: Noam Yorav-Raphael
- [ ] @obiwanus
- [ ] @orivej
- [ ] @pjvandehaar
- [ ] @sizmailov
- [ ] @Socialery <- no full name
- [ ] @staffanm
- [ ] @stonebig <- no full name
- [ ] @tacaswell
- [ ] @tiagocoutinho
- [x] 1 @casperdcl: Casper O. da Costa-Luis, ORCID 0000-0002-7211-1557
- [x] 2 @lrq3000: Stephen Karl Larroque, ORCID 0000-0002-6248-0957
- [x] 5 @hadim: Hadrien Mary, ORCID 0000-0001-8199-5932
- [x] 8 @altendky: Kyle Altendorf
- [x] 11 @kmike: Mikhail Korobov, affiliation Scrapinghub Inc.
- [x] 14 @MPagel: Matthew D. Pagel, affiliation University of California, Davis
- [x] 16 @hugovk: Hugo van Kemenade, ORCID 0000-0001-5715-8632
- [x] 17 @igorljubuncic: Igor Ljubuncic, affiliation Snap Advocacy team
- [x] 18 @uadnan: Adnan Umer
- [x] 20 @jnhansen: Johannes Hansen
- [x] 21 @RunOrVeith: Veith Röthlingshöfer, ORCID 0000-0002-1824-3153
- [x] 26 @FichteFoll: only add if GitHub handles are fine, otherwise don't
- [x] 28 @mihaild: Mikhail Dektyarev
- [x] 29 @RedBug312: Chung-Kai Hung
- [x] 32 @aplavin: Alexander Plavin, ORCID 0000-0003-2914-8554
- [x] 37 @kcwu: Kuang-che Wu, affiliation Google Inc
- [x] 41 @dwhswenson: David W. H. Swenson, affiliation École Normale Supérieure de Lyon
- [x] 43 @Separius: Sepehr Sameni
- [x] 45 @cheng10: Cheng Chen
- [x] 52 @cancan101: Alex Rothberg
- [x] 54 @carlinmack: Carlin MacKenzie, ORCID 0000-0002-9300-0741
- [x] 57 @EdwardBetts: Edward Betts
- [x] 70 @davidbau: David Bau
- [x] 72 @arkottke: Albert Kottke
- [x] 73 @Berserker66: Fabian Dill
- [x] 76 @shirish93: Shirish Pokharel

## asked to be removed

- [x] 6 @mjstevens777
- [x] 13 @mbargull
- [x] 19 @immerrr
- [x] 23 @mikekutzma
- [x] 24 @yarikoptic
- [x] 25 @anntzer
- [x] 27 @julienc91
- [x] 33 @songololo
- [x] 36 @maxnordlund
- [x] 39 @willtrnr
- [x] 43 @jcea
- [x] 47 @darindf
- [x] 49 @robertzk
- [x] 53 @toddrme2178
- [x] 56 @andportnoy
- [x] 57 @arunpersaud
- [x] 58 @danking
- [x] 61 @deeenes
- [x] 66 @f0k
- [x] 67 @JaviMerino
- [x] 69 @JoshKarpel
- [x] 69 @volans-
- [x] 71 @elsonidoq
- [x] 72 @rlmaers
- [x] 76 @to266
- [x] 77 @zed

casperdcl committed:
update paper auth list

- fixes #887

casperdcl committed:
more requests

casperdcl committed:
more ref updates

volans- commented:
@casperdcl my single contribution was just to fix a typo in documentation and docstrings. Feel free to include or exclude me based on where you draw the line of co-authoring. I tend to think that I don't deserve it.

casperdcl committed:
more auth updates

hugovk committed:
Spell project names as they spell them

hugovk committed:
Update impact stats for March 2020

casperdcl committed:
more auth updates

hugovk committed:
Update Arabic spelling

- closes #907

dwhswenson commented on examples/paper.md:
```diff
@@ -147,6 +144,8 @@ affiliations:
    index: 4
  - name: Google Inc.
    index: 5
+ - name: Université de Lyon
```
If going with the short version, it should be "École Normale Supérieure de Lyon" for me

Thanks!

casperdcl committed:
udpate affilition

casperdcl committed:
minor status update

hugovk committed:
Update years for updated data

casperdcl committed:
minor affiliation reorder

lrq3000 commented:
Checking now, sorry for the delay, things are kind of crazy right now..

lrq3000 commented:
@casperdcl Thank you very much for building this list! That's awesome!

To finish this up I propose two things to fix the last two issues:

* About people who did not reply: I think we should not put them as co-authors as they did not give their explicit consent, it's not super ethical. I'm sad some of the important contributors could not reply, but we can't choose for them. What I propose is to add them in the Acknowledgements section. And we could also add the ones who refused to be co-authors, that seems fair and ethical to me.

* About co-author list, I agree to follow `git fame -wMC` except that the first 3 authors should be Casper da Costa-Luis, Stephen Karl Larroque and Noam Yorav-Raphael, as there was a clear consensus by nearly all commentators on these co-authors and order.

jcea commented:
I don't even remember what I patched in tqdm, so being coauthor is flattering but probably excessive. Being in an Acknowledgements section would be less embarrassing :).

kmike commented:
hey! I'm fine with having me on a list, with Scrapinghub Inc. affiliation if that's not too much work. Thanks!

f0k commented:
Sorry, only saw this now. [My contribution](https://github.com/tqdm/tqdm/pull/713) wasn't significant, so I don't want to be a co-author, but I wouldn't object being listed in the Acknowledgements if you see fit.

casperdcl committed:
more auth updates

casperdcl commented:
Please post discussions and requests in the issue #887. @lrq3000 note that https://github.com/tqdm/tqdm/pull/905#issuecomment-599143376 is the exact opposite of what was agreed. If you would like to re-open the entire debate from the start please do so at #887.

lrq3000 commented on .mailmap:
```diff
@@ -1,17 +1,20 @@
-Max Nordlund <max.nordlund@gmail.com>
+James Lu <CrazyPython@users.noreply.github.com>
+Hugo van Kemenade <hugovk@users.noreply.github.com>
 Julien Chaumont <git@julienc.io>
```
asked to be removed

lrq3000 commented on .mailmap:
```diff
@@ -1,17 +1,20 @@
-Sergei Izmailov <sergei.a.izmailov@gmail.com>
-James Lu <CrazyPython@users.noreply.github.com>
+Alexander Plavin <alexander@plav.in>
+Max Nordlund <max.nordlund@gmail.com>
```
Max Nordlund wants to be removed

lrq3000 commented on .mailmap:
```diff
@@ -1,17 +1,20 @@
-James Lu <CrazyPython@users.noreply.github.com>
+Alexander Plavin <alexander@plav.in>
+Max Nordlund <max.nordlund@gmail.com>
 Darin <darin@omegasoft.org> darindf
```
Darindf asked to be removed

lrq3000 commented on .mailmap:
```diff
@@ -1,17 +1,20 @@
+Todd <toddrme2178@gmail.com>
 Pablo Zivic <elsonidoq@gmail.com>
-Albert Kottke <albert.kottke@gmail.com>
 Riccardo Coccioli <volans-@users.noreply.github.com>
```
Riccardo Coccioli asked to be removed

lrq3000 commented on .mailmap:
```diff
@@ -1,17 +1,20 @@
 Darin <darin@omegasoft.org> darindf
+Sergei Izmailov <sergei.a.izmailov@gmail.com>
+Todd <toddrme2178@gmail.com>
 Pablo Zivic <elsonidoq@gmail.com>
```
Pablo Zivic (elsonidoq) asked to be removed

lrq3000 commented:
@casperdcl Thank you very much for building this PR, it's awesome, great work!

Yes you are right, I have changed my mind, as indeed we initially decided to be as inclusive as possible. This is a tough call, both decisions have ethical pros and cons. But I agree we should stick to the initial consensus and goal of being inclusive, so to add by default as co-authors.

I have just one change suggestion: it seems there was a consensus for Noam to be 3rd author, so I would suggest to place him as third.

Also I'm wondering why is his affiliation "independent" instead of "tqdm developers"? Is it because you considered the "tqdm developers" as the github organization that was formed after the fork? If that's the case, I understand the idea, but I would lean on changing the affiliation to "tqdm developers" for him, as I am not sure it's meaningful to make such a distinction in the paper between the author of the initial repo and the subsequent main devs.

lrq3000 committed:
Paper: fix spelling Jupyter HTML notebooks -> Jupyter Notebook

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 reviewed (changes requested):
Check Noam position (3rd?) and affiliation, otherwise ready for merge.

casperdcl committed:
catch-all tqdm developers author

Seems to be an approach for other JOSS papers

casperdcl committed:
Merge pull request #910 from hugovk/project-names

casperdcl committed:
update all stats again

casperdcl committed:
Merge branch 'update-impact-stats' into paper

- closes #911

casperdcl committed:
update order

as per `git fame -wMC` using

- https://github.com/casperdcl/git-fame/commit/274df71ec0705568bbd46cda8bb7f540007c3e82
- https://github.com/tqdm/tqdm/commit/1e0bb74db632c10643ff5c3e8ae699bb99118c38

casperdcl commented:
Ok I think this is ready to merge now.

@lrq3000, concerning the only remaining point about ordering authors:

I don't view the list as being an author list at all since it completely violates most people's definition of authorship (including the IEEE definition) as discussed. I view it as a contributor list, i.e. people who have contributed in some way to the project (as opposed to the JOSS paper). This has the added benefit of being inclusive.

Objective ordering metrics are flawed but still infinitely better than arguing (vis https://github.com/psf/black philosophy). Therefore in terms of ordering, it makes most sense to me to leave things as they are.

However, I am willing to accept anything which can be objectively programatically calculated.

Examples:

- order based on number of SLoC
- order based on number of commits
- order based on `numpy.pi * (author.last_commit_time - author.first_commit_time) ** 2 / len(author.name)`

I am not willing to accept anything based on any human's subjective "feelings."

Examples:

- order based on `first_commit_time` except put some person first (@casperdcl) because they wrote most of the code/the entire paper/website/contributed the longest/some humans have subjective "feelings" that they should be first
- order based on number of commits except put some person third (@noamraph) because they uploaded 150 lines of code 6.5 years ago/some humans have subjective "feelings" that they should be third

noamraph commented:
Hi,

Since:

1. It seems that everyone except @casperdcl agrees that I should be third, or doesn't mind,
2. I would pretty much like to be third,
3. It seems to me that there's no point in arguing with @casperdcl, who defined my contribution as "150 lines of code 6.5 years ago",

I therefore retract my approval for this paper, unless I am listed as third author.

My hope is that the JOSS won't agree to publish this paper without my approval, and thus will require @casperdcl to accept what everyone else agrees to. If I'm wrong, that's OK too.

If I'm wrong about point 1, please correct me.

Cheers,
Noam

casperdcl commented:
Ok posted https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-640165990

noamraph commented:
Just so it will be clear: does anyone except for @casperdcl object to @lrq3000 suggestion?
 > follow `git fame -wMC` except that the first 3 authors who should be Casper da Costa-Luis, Stephen Karl Larroque and Noam Yorav-Raphael, as there was a clear consensus by nearly all commentators on these co-authors and order."

If that's the case, I don't really understand why @casperdcl would stop publishing this paper, if everyone agrees he should be first author, and there's a consensus among all others about the order of the rest of the authors.

lrq3000 commented:
> However, I am willing to accept anything which can be objectively programatically calculated.
> [...]
> I am not willing to accept anything based on any human's subjective "feelings."

@casperdcl I think there is a misunderstanding here, because Noam's contribution can certainly be objectively quantified.

Noam's contribution is not just that he created the package: he laid the foundation on which `tqdm` is still based up to this day. Which is quantifiable.

I ran a little experiment using [mfilecompare](https://www.mathworks.com/matlabcentral/fileexchange/29781-mfilecompare), a diff comparison tool usually used for plagiarism but which I use here to quantify the inclusion/remanence of the Noam's contributions. It was made for MATLAB but it works well for other languages too (it just reduces to a diff tool under the hood). I included in the comparison:

* [noam's tqdm repository](https://github.com/noamraph/tqdm) (which also includes a few commits by arkottke and JackMC)
* 3 files from current tqdm: std.py, notebook.py, gui.py - the latter two serving as a functional control (although some overlap is still expected since they were derived from std.py itself derived from noam's tqdm.py, but the overlap should be a lot less than between std.py and noam's tqdm.py).
* pbar.py from [progressbar2](https://github.com/WoLpH/python-progressbar), which serves as a baseline control.

Here is the result:

![content-comparison](https://user-images.githubusercontent.com/1118942/83965734-2a92cf00-a8b6-11ea-94fa-da95747dad3f.png)

More precisely, here are some of the most interesting numbers (using `x C y` meaning x includes as much content from y):
* **std C noam: 31%**
* std C pbar (progressbar2): 7%
* pbar C noam: 6%
* gui C noam: 9%
* notebook C noam: 4%

In other words, the software finds that even nowadays, std.py (the core of tqdm) includes approximately 31% of noam's tqdm.py, which is way beyond chance or just noise induced by just using the same language (pbar) or similar naming conventions (gui and notebook). And that's just a rough approximation, because when looking manually, it's clear nearly everything that was laid is still used.

This quantification just illustrates the significant and obvious contributions of Noam: the core mechanics of tqdm were made by him. Else, the current tqdm would either not exist or would be based on another project. By looking at the qualitative nature of Noam's contributions, it becomes clear why, as the original tqdm already included:

* trange as an easy wrapper for tqdm and range
* a decoupled iterations and time memoization scheme to reduce the overhead of time.time()
* checking the counter first to reduce the overhead of time.time()
* tqdm working as a wrapper to any iterable
* a decoupled and hence configurable formatter
* ETA and number of iterations per second estimation

And probably more I have missed but I am not trying to be exhaustive here. But that's why Noam's code is still prevalent in the current tqdm's codebase, and I would bet he's among or is the top one for maintained code throughout the codebase history (I would love to see a tool to quantify that but that doesn't seem to exist). Because yes he didn't write much compared to others, but his contributions were so critically useful they were kept all along. That's the reason I think he deserves to be third author, and not just because he is the original creator (although of course both are linked).

I think this is a great illustration of what gkthiruvathukal [wrote](https://github.com/tqdm/tqdm/issues/887#issuecomment-578523686):

> There are multiple ways to make an impact, and sometimes an enormous amount of impact can be made without the singular focus on LOC (or commits).

lrq3000 commented:
@casperdcl Also thank you very much for attending to the remaining issues that needed some care, and for your work on both tqdm and the paper.

I hope the quantification I have given above can convince you of the objectivity and adequacy of placing Noam as 3rd author. We should also not forget that there seems to be a consensus (but not unanimity), which subjectivity is debatable of course, but it still is something important to consider.

casperdcl commented:
I'd like to clarify that I am not in any way belittling anyone's contributions. I absolutely believe everyone (Noam included) has made important, significant contributions to the code. I have made numerous extreme efforts to highlight as many contributions as possible without anyone asking me to do so. This includes developing https://github.com/casperdcl/git-fame, and adding links to numerous contributor pages at https://github.com/tqdm/tqdm/#contributions, including even explicitly mentioning Stephen and Noam. I have repeatedly stated that all contributors are important, cared for, and deserve any and all attention.

However in the absence of a JOSS definition of paper authorship, I can only go by definitions of other respected organisations. My original approach was to follow IEEE (authors must write the paper and do not have to seek permission or invite co-authorship from anyone to write said paper), however I am happy to use a more inclusive Zenodo approach - i.e. automate (based on e.g. #commits or SLoC).

arfon commented:
> However in the absence of a JOSS definition of paper authorship, I can only go by definitions of other respected organisations. 

I believe there is a third way here which is that if the group is able to make a consensus agreement about authorship then ordering then that is acceptable too.

Speaking for my own research discipline, a common approach is for the group to agree the first _n_ authors (usually ~2-4 people), and then for the rest of the authors to be ordered alphabetically. This allows authors with significant contributions (yes, I realize this phrasing is subjective) to be first, and all others to be ordered in some automated fashion.

lrq3000 commented:
@casperdcl I appreciate the spontaneous efforts you have made to indeed mention on the readme all the authors in the tqdm project. I did not mean to intend that Noam's contribution was disregarded, but I understood your point that we should avoid a subjectively defined order, and hence I provided a quantification (of surviving textual content ≥ ~30% of original tqdm pre-fork).

This quantification does not seem to be incompatible with a combination with the metric we used before with git-fame, I don't think we need to stick to only one measure, furthermore given the known limitations (eg, a surviving line of code is quite tricky to define - one example: is a variable name change enough to consider the line changed?).

There is also the consensus, which is not a formally defined thing, but it is neither completely subjective.

To be honest Casper, I do not understand why and where is the issue. This doesn't impact your position as of course you merit the first co-authorship, but everyone - except you - who voiced an opinion spontaneously agreed on this order for the first 3 authors (and for the rest of authors, they all agreed they didn't care about the order), so I am not sure why this isn't already resolved? If this is a problem of metrics, please clarify why you think we should stick with only one metric and disregard others?

lrq3000 commented:
@hadim sorry it seems I unassigned you from this issue but I didn't do any particular action so that's weird. And I can't reassign you, you're not in the list of people I can reassign? :-/

kmike commented:
My 2c: measuring contribution by LoC or other "objective", automated metrics is misleading. Imagine your employer to implement such evaluation against you. For me it is obvious that @noamraph, as an original author, is among the top contributors to the project, and should be included in top 3 regardless of what "objective" metrics say.

lrq3000 committed:
Move authors beyond 3rd co-author to contributors at the top of document

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
add acknowledgements to explain what is the Contributors field at the top + change affiliation (it should be an institution, tqdm maintainers is not one unfortunately)

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
add all code contributors names per https://github.com/tqdm/tqdm/pull/905#issuecomment-599143376 + alphabetical order and duplicates removal

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
Clarify general problematic in intro

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
clarify that alphabetical order is used for contributors list

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
Add a manual example and 2 features + extend performance monitoring

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: copyediting

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: same style for 2nd example as 1st

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: copyediting

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: copyediting

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: copyediting

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: remove Yahoo Answers reference and tqdm abbreviation in spanish

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: update order of contributors by contributions via `git-fame`

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 commented:
Dear @casperdcl, dear @noamraph, dear everyone,

As you may have seen, Arfon (from the JOSS editors) made a [new statement](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660692001) on which we should build upon to reach a conclusion in this issue before the 2nd August 2020.

I rewrote a draft in my [cloned repo](https://github.com/lrq3000/tqdm/blob/paper/examples/paper.md) which can be previewed here: https://res.cloudinary.com/hju22ue2k/image/upload/v1595281024/dgrdymzuzyruunhwms02.pdf

This draft includes the following changes that should meet Arfon and our consensual criteria:
* 3 co-authors in the following order: @casperdcl, me, @noamraph .
* All other contributors to the tqdm project listed as "Contributors" at the top of the document, just below the list of the paper's co-authors, order with `git fame -wMC`. I double checked the names to track down a few more real names and ensure there was no duplicate. I added back all contributors including those who did not reply or opted out of co-authorship. Since some contributors who opted out of co-authorship mentioned they would nevertheless agree to be acknowledged, I suggested [in the past](https://github.com/tqdm/tqdm/pull/905#issuecomment-599143376) that all contributors could be acknowledged whether they co-opted out of authorship. Since we will now have 3 authors, I think it's a good opportunity to list all contributors.
* Affiliation was changed to Independent for all 3 co-authors, since we do not (or did not at the time of contributions) belong to an institution.
* Rewritten problematization part of the introduction and added an example and a few features. Although content edition wasn't planned in this issue, if all 3 co-authors agree I think JOSS will see no objection.
* Removed the spanish abbrevation of tqdm, the reference (Yahoo Answers) is not adequate for a publication. I could not find an alternative source unfortunately, so I decided to remove this detail.

Please let me know your thoughts/feedback/suggested changes @casperdcl and @noamraph. If it looks good to you I'll merge my branch into this PR.

lrq3000 committed:
paper: try to fix syntax highlighting of output of examples

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: copyediting

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
paper: arabic characters unsupported by the JOSS generator

Signed-off-by: Stephen L. <lrq3000@gmail.com>

noamraph commented:
@lrq3000 it looks fine to me. Thanks!

hugovk committed:
Remove unused Yahoo Answers! reference

hugovk committed:
Use plain English 'use' instead of 'leverage'

hugovk committed:
Use plain code block for output to avoid incorrect syntax colouring

hugovk committed:
Software is a mass noun and doesn't take an indefinite article

hugovk committed:
Fix Flake8

hugovk commented:
Here's the diff: https://github.com/tqdm/tqdm/compare/paper...lrq3000:paper

See also https://github.com/lrq3000/tqdm/pull/1 for some minor updates to that draft.

hugovk committed:
Use console formatting for examples with $ prompts and command output

lrq3000 committed:
Merge pull request #1 from hugovk/paper

Update the paper

lrq3000 commented:
Thank you very much @hugovk for reviewing and the fixes :-)

I think it's time to merge the changes in this PR here. I'll do that now unless someone makes an objection.

lrq3000 commented:
Thank you @noamraph for your fast reply.

@casperdcl could you please have a look soon? I think it would be good if we can be early on the deadline as there are 4 reviewers assigned so they need to have time to read the paper. Thank you in advance :-)

lrq3000 commented:
@casperdcl The deadline given by JOSS editors is soon (2nd of August, this sunday), could you please share your feedback as soon as possible?

lrq3000 commented:
@casperdcl Your last reply on the matter was 9 days ago. Could you please reply or at least just give us an update? This issue has gone on for too long. If you don't reply by Friday included, on Saturday I will notify JOSS editors of our failure to reach an agreement and that the paper should be retracted.

casperdcl commented:
Not sure what's going on.
To repeat, I (@casperdcl) *insist* on either of two definitions of authorship:

- IEEE (paper-centric; paper authors and subject/software authors are unrelated; no consent is required in either direction)
- Zenodo/GitHub (software-centric; author lists are automatically generated; there are no paper authors; there are only software authors)

noting that I have made consistent extreme and exceptional efforts to follow the first definition in the journal paper; and the second definition in the software documentation & revised journal paper.

Meanwhile, JOSS (CC @arfon) *insists* on:

- Consensus ("authors themselves are responsible for deciding [...] authorship")

noting that it is an independent journal free to insist on whatever it wants.

All three definitions are hopelessly flawed. However, their relative merits and weaknesses are NOT to be debated here. This is NOT a forum to discuss definitions.
This is solely a forum to acknowledge that these definitions are incompatible (for future reference).

(On an unrelated note, I'm not sure why @lrq3000 is re-discussing a fourth proposal expressly rejected months ago.)

casperdcl commented:
wait. what. hang on... did you just merge this in without approval and then threaten me with my own words? This is a mess.

altendky commented:
For what it's worth, @casperdcl, you are not being inclusive and respectful.  Everyone else that I've seen express an opinion is happy with giving you primary credit etc, etc.

Sad, but oh well.

lrq3000 commented:
Wow, that's impressive, i didn't expect you @casperdcl to go this far.

For info, Casper just revoked my membership as a tqdm maintainer and hence all my rights on the tqdm repo. Apparently, updating a PR branch was trespassing my rights as an assigned member on this PR and the 2nd top contributor ever on the repo.
Let's be honest Casper, that's just an excuse for kicking me out. PR branches are just drafts, you know any commit can just be cherry picked or rollbacked. When you changed the licensing of your contributions without discussing beforehand and hid in a myriad of changed lines of code so that nobody noticed until a few weeks later when a user opened a PR, that was disrespectful and worthy of a ban, but i didn't do it. I do not regret what I did, but I regret what you did here.

Before continuing further, let's clarify something *once again*: there is no reason you should be the sole paper author just because you wrote it, because anybody can write a paper about tqdm. What if I write a paper on my own about tqdm and submit it to another journal without mentioning nor notifying you, would that be fine to you? If Noam (the creator of tqdm) or any other contributor did that, would that be fine? Why would it be fine for you to do that?

That's why ethics and consensual practices exist about paper authorship: by inviting those who contributed to the project whether they want to co-author the paper, this solve the issue of defining who will be the authors. But you did not do that. So we can't know who would have been authors. Writing a paper solo without ever inviting other contributors beforehand is not part of any academic practice I know of (at least in reputable institutions). This root ethical issue and the consistent failure to address it is not what I would call a "consistent extreme and exceptional efforts".

There is no single definition of authorship, and JOSS editors made clear the condition for the publication to be maintained: reach a consensus. You're just trying to hide behind definitions to avoid saying clearly that you simply flat out refuse to give any meaningful credit for Noam's contribution. But remember this is the root cause that raised this issue, and it seems you never really intended to listen to any other voice on the matter. I never, *ever*, saw any work where it was considered ethical to not even mention its original creator.

> (On an unrelated note, I'm not sure why @lrq3000 is re-discussing a fourth proposal expressly rejected months ago.)

This is nonsensical and looks like a [reversal of roles](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660556781), but I'll try to answer from what I can interpret. I am making no new proposition (except for the content but I don't think you meant that), for the authorship and contributors I am following the suggestions by JOSS editors following the roadblock from your refusal to reach a solution by consensus. This approach (of having all contributors acknowledged) was never rejected, it has always been an open possibility depending on what would be our possibilities in the end.

Finally, no, that's not threatening, it's a reminder that not giving just at least an update until after 11 days out of the 14 days we have for this last deadline is simply disrespectful of your collaborators. You did not even reply to @arfon 's [statement](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660692001)! I don't want us to disrespect JOSS editors' time anymore, so i do intend to fulfill the deadline, whatever is the result.

So either we reach a consensus (me, @casperdcl and @noamraph) by Sunday, or I'll notify JOSS editors of our failure to reach an agreement and hence the paper's retraction per the stated conditions.

casperdcl commented:
# Background

I submitted a paper I wrote (about some software I maintain) to a journal.
A reviewer flagged that an arguably significant ex-contributor might be interested.
Despite 2 years of not responding to my `@`mentions and emails, I still offer co-authorship in the hope of incentivising continued contributions.
There was no response. The paper was later accepted and published.

Eight months elapse, and the ex-contributor suddenly gets in contact, demanding co-authorship.
This seems unreasonable to me, but "if the journal is fine with it, whatever" I think.
The ex-contributor then demands acknowledgement of all (over 50) software contributors.
This also seems unreasonable, but "if the journal is fine with it, whatever" I think, again.
I've even made a tool which could help collect this massive author list - or we could rely on respectable alternative sources such as Zenodo or GitHub.
All goes well, after a considerable amount of effort on my part.

The ex-contributor then makes a third demand - to manually reorder the list of authors based on personal preferences without any foundation whatsoever on any metric.

Misrepresentation is fraud, and fraud is a crime, no matter who thinks it fine.

Furthermore, against all credulity, the ex-contributor demands an even worse alternative - remove all but 3 of the authors.

The ex-contributor then goes on to ignore my explicit refusal, and over the course of the next few weeks, removes all but 3 authors.

The ex-contributor then merges in these changes without approval. **The first commit the ex-contributor has bothered to make in 3 years is to remove everyone I painstakingly added to a paper's author list.**

The ex-contributor then hypocritically insists that I respond instantly (within hours or days), and furthermore threatens that they somehow have the authority to give instructions to a journal.

# Summary

1. You may inform a journal that you think a paper which you did not write should be retracted. However you may not presume the authority to actually decide. If the journal does decide to retract the paper, it is not because you wanted it. It is because of the journal's own policies.
2. It is shameful that anyone even needs to explanation how incredibly illegal (not to mention immoral, unethical, and downright stupid) this 3-author-followed-by-50-contributor idea is.

- Using an "exclusive" definition of authorship, a paper author is the person who wrote the paper.
- Using an "inclusive" definition of authorship, a paper does not have an author, and we simply list all of the contributors as "authors."
- Using an absolutely evil, wrong "definition" of authorship, only 3 contributors will be listed as "authors," and the rest immediately below as "contributors." This implies that exactly 3 people have authored something while supervising and directing the rest.

The rejected proposal implies that 3 FOSS contributors are somehow special; more valuable, important, or essential compared to all others. So important that only they will be awarded a different title: "author," which legally means that only they "own" some work, whether the "work" is the paper or the software or both. Any possible interpretation is an abominable lie. The software is "owned" by many more people under a FOSS licence, and the paper is "owned" solely by me in the sense that I have sole copyright over the words in the paper I wrote, submitted, and licensed out to a journal to publish under CC-BY-4.0.

The rejected proposal tells previous, current, and future contributors that they will be arbitrarily discredited from their software contributions by being non-consensually split into separate groups of different names (in this case "author" and "contributor"). This is not in any way compatible with the spirit of FOSS and will not encourage future contributions, and I thus wholeheartedly reject it.

----

Stephen (@lrq3000), you are still publicly listed as a member of the `tqdm` organisation.
If, after 3 years of not contributing to the project despite repeated pleas, you decide to resume contributions in a manner which requires re-instatement of admin privileges, I would be more than happy to oblige.

casperdcl commented:
> What if I write a paper on my own about tqdm and submit it to another journal without mentioning nor notifying you, would that be fine to you? If Noam (the creator of tqdm) or any other contributor did that, would that be fine? Why would it be fine for you to do that?

Yes, @lrq3000, this is explicitly allowed by IEEE rules and something I explicitly already addressed months ago. Of course it would be fine. Writing a paper about clouds doesn't mean you assert ownership of clouds.

altendky commented:
Your comments tell existing and potential contributors, authors or otherwise, that they will not be respected.  That you will be the ultimate decider of morality let alone everything else.  That all others involved in tqdm are both illegal and evil.  That is what is you are saying, @casperdcl.

Despite this, _everyone_ else wants to give you primary credit and agrees with the proposal.  Anyways, sounds like time for you to report us all to the authorities for our conspiracy to commit illegal acts.

(Seriously?  I can write this without any amount of hyperbole being involved?  This is truly amazing.)

altendky commented:
> I value all contributors.

Then accept the will of _all_ other contributors regarding this authorship and acknowledgement topic instead of insulting us.

lrq3000 commented:
Well, at least we got some clarification from @casperdcl 's post, from its intro ("an arguably significant ex-contributor") and his consideration that recognizing Noam's major contribution in the creation of tqdm constitutes a "fraud". This last bit is simply outrageous, but at least he finally states he has no intention to recognize Noam's contribution. If that was not clear enough before, all contributors but him are arguably insignificant.

The whole recap above by @casperdcl is a travesty of truth. He did not invite me to co-author the paper, he was [**compelled to**](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-469208417) because it was too blatant even for external reviewers such as JOSS editors to not notice that a single author for this paper was odd. I did not reply at the time because I did not want to chime in just because the editors required @casperdl to contact me, as clearly if he wanted me on board, he would have asked before. Remember that I raised concerns because months after when I stumbled on the publication, I was appalled to see that the paper did not even mention anyone but Casper, not even the original creator Noam. Not inviting fellow contributors to co-author is one (bad) thing, but **erasing them from history** by not even acknowledging them is another that I could not imagine. This is the definition of a fraud in academic works. So yes, I should have checked the paper before, but I trusted that @casperdcl 's experience in academia would not allow for this kind of aberration.

This erasing from history is not a single instance, it's repeated all over the place: from the [Readme copyrighted solely by Casper](https://tqdm.github.io/) to the [PyData 2019 talk](https://youtu.be/kM3G35yhyDw?t=340) claiming that "development began in 2015" (erasing Noam's contributions in 2013 to 2014), and the [History page on the Wiki](https://github.com/tqdm/tqdm/wiki/History/7053db9cdf7ab3b1017d3f584755c380aba2bf44) which does the same and where nobody but Casper gets mentioned, a one-man history in other words.

Note that I do not know Noam, I never talked to him before he replied to this issue. I simply respected his work, as it's only normal he was **at least** mentioned as the original creator of tqdm.

> Yes, @lrq3000, this is explicitly allowed by IEEE rules and something I explicitly already addressed months ago. Of course it would be fine. Writing a paper about clouds doesn't mean you assert ownership of clouds.

That's completely incorrect, and this raises questions on your understand of how academic works are done. Nobody writes (academically) about clouds, they write about models or observations about clouds. If these models or data were made/collected as the result of a collaborative effort, then all contributors have to be invited to co-author or AT LEAST be acknowledged, neither of which you did. Since you like to reference IEEE, their manual explicitly states this can be considered a case of [plagiarism](https://pspb.ieee.org/images/files/files/opsmanual.pdf) (Subsection 8.2.1-B.7).

Lastly, I want to clarify that I am perfectly fine with having all contributors as co-authors, I only implemented [the suggestion by JOSS editors](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660692001). Having all contributors listed just below the co-authors as suggested by Arfon appeared to be an elegant solution to me to fix all issues at once, including Casper's legal concerns. But I'm of course very fine with either, as long as all contributions are adequately credited, including Noam (which means 3rd co-author per consensus). It's hence very hypocritical and of very bad taste from @casperdcl to try to tarnish me with these fabrications (among other things):

> Furthermore, against all credulity, the ex-contributor demands an even worse alternative - remove all but 3 of the authors.
> The ex-contributor then goes on to ignore my explicit refusal, and over the course of the next few weeks, removes all but 3 authors.

@casperdcl proposed out of nowhere that [ALL contributors but me and him be removed from authorship](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660536262), which I of course refused. 

Given that @casperdcl has clearly no intention in *properly* crediting Noam, and I doubt anybody else given [his last proposition](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660536262), as he considers doing so would be a "fraud" (whereas not crediting anybody else was fine for him), I guess we should just wrap this up and admit to JOSS that we failed. It's unfortunate, but that's @casperdcl 's choice. Everybody respects his work, but the reciprocal isn't true.

hadim commented:
Two things:

- @casperdcl: next time you talk or write about tqdm and its history, I ask you to explicitly cite me as a contributor of the project. And do the same for ALL of them.

- this paper shouldn't published given your last answer @casperdcl.

lrq3000 commented:
It seems we are in a dire state, but I think it's worth to try one last time nevertheless.

After thinking with some hindsight, I am convinced Arfon's suggestions in his last statement were based on Casper's suggestion. This includes the move of most contributors into a dedicated "Contributors" line just below the authors line, at the top of the document. I implemented it since I thought it would save time for Casper who anyway expressed his wishes to see this kind of plan in [his last proposition](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660536262) just before Arfon's statement. Removing my writing rights for this is I hope a misunderstanding, although I fear it's not, but let's get past that.

This made me realize that since the beginning, everyone tried their best to meet most if not all of your suggestions @casperdcl. The only real disagreement you have is about Noam's co-authorship and position. So let's focus on this from now on.

The last statement by Arfon changed just one thing: @noamraph agreement is now necessary to conclude this issue. Since the beginning, the main crux is to adequately credit his work as the creator of tqdm. It's also an opportunity to credit other contributors, and it seems we all agree on this part. But the issue remains about Noam's crediting, and addressing it is not a favor, it's a requirement, JOSS editors recognized that since we are where we are now.

Me and @noamraph both agree he should be 3rd coauthor. If you fear that could be a "fraud" per academic practices, I can assure you it would not be and I am willing to take full responsibility, and the fact that not only other contributors but also JOSS editors see no issue makes me feel confident this is not an issue.

Would you @casperdcl agree with having Noam as 3rd coauthor? If yes, we can further work on completing this issue before the deadline, if not and hence fail to reach an agreement, this means the paper will get retracted unfortunately.

lrq3000 commented:
Dear JOSS editors, dear collaborators,

It is with regrets that i must acknowledge that we failed to reach an agreement despite our collective efforts, and it is unlikely any agreement can be reached. We respect the [editors decision of a retraction](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-667670005) per the [previously stated conditions](https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-660692001). 

On behalf of all the parties involved, I would like to thank @arfon and the JOSS editorial team for their careful and helpful guidance. Please accept our sincere apologies for the effort and time it took to reach this unfortunate conclusion.

Although the result is certainly disappointing for everyone, and some bad things were said, there also were very heartwarming altruistic moments. Thank you to everyone who had the courage to participate. I am honored to have worked with you on this project.

Let's now put this issue to rest. (including #887).

hadim commented:
Sad but I support @lrq3000 statement.

hadim commented:
For the record the paper has been retracted:

- https://joss.theoj.org/papers/10.21105/joss.01277
- https://github.com/openjournals/joss-reviews/issues/1277#issuecomment-667670005

![image](https://user-images.githubusercontent.com/528003/89240191-65b93200-d5c9-11ea-846d-6c8cb5a28d99.png)
